File: closure/python/closure.py
class Player:
    def __init__(self, name, nation, age):
        self._name = name
        self._nation = nation
        self._age = age

    def get_name(self):
        return self._name
    
    def get_nation(self):
        return self._nation
    
    def get_age(self):
        return self._age
    
    def get_value_by_property(self, prop):
        switcher = {
            "name": self.get_name(),
            "nation": self.get_nation(),
            "age": self.get_age()
        }

        return switcher.get(prop)

def distinct_by(prop): 
    store = set()

    def has_score(player):
        value = player.get_value_by_property(prop)
        
        if value == None: 
            return False

        if value in store:
            return False

        store.add(value)

        return True

    return has_score

def get_nation(player):
    return player.get_nation()

players = [
    Player("messi", "argentina", 33),
    Player("ronaldo", "portugal", 35),
    Player("son", "korea", 29),
    Player("park", "korea", 39)
]

player_filtered_by_nation = list(filter(distinct_by("nation"), players))
nations = list(map(get_nation, player_filtered_by_nation))

print(nations)

File: closure/golang/closure.go
package main

import (
	"fmt"
)

type Player struct {
	name	string
	nation	string
	age		int
}

func (p Player) GetValueByProperty(property string) interface{} {
	switch property {
		case "name": return p.name
		case "nation": return p.nation
		case "age": return p.age
		default: return nil
	}
}

func (p Player) getNation() string {
	return p.nation
}

var players = []Player {
	Player{
		name: "messi",
		nation: "argentina",
		age: 33,
	},
	Player{
		name: "ronaldo",
		nation: "portugal",
		age: 35,
	},
	Player{
		name: "son",
		nation: "korea",
		age: 29,
	},
	Player{
		name: "park",
		nation: "korea",
		age: 39,
	},
}

func Filter(players []Player, callback func(Player) bool) []Player {
	result := []Player{}

	for _, v := range players {
		if (callback (v)) {
			result = append(result, v)
		}
	}

	return result
}

func Map(players []Player, callback func(Player) string) []string {
	result := []string{}

	for _, v := range players {
		result = append(result, callback(v))
	}

	return result
} 

func DistinctBy(key string) func(player Player) bool {
	store := map[interface{}]interface{}{}
	return func(player Player) bool {
		value := player.GetValueByProperty(key)
		
		if (value == nil) {
			return false
		}

		if (store[value] == true) {
			return false
		}

		store[value] = true		

		return true
	}
}

func main() {
	playersFilteredByNation := Filter(players, DistinctBy("nation"))
	nations := Map(playersFilteredByNation, func(player Player) string {
		return player.getNation()
	})

	fmt.Println(nations)
}